File: src/exercises/course_schedule_4/python/solution.py
def checkIfPrerequisite(numCourses: int, prerequisites: list[list[int]], queries: list[list[int]]) -> list[bool]:
    adj = [set() for _ in range(numCourses)]
    indegree = [0] * numCourses
    isPrereq = [set() for _ in range(numCourses)]
    
    for pre, crs in prerequisites:
        adj[pre].add(crs)
        indegree[crs] += 1
    
    q = deque([i for i in range(numCourses) if indegree[i] == 0])
    
    while q:
        node = q.popleft()
        for neighbor in adj[node]:
            isPrereq[neighbor].add(node)
            isPrereq[neighbor].update(isPrereq[node])
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                q.append(neighbor)
    
    return [u in isPrereq[v] for u, v in queries]

File: src/exercises/course_schedule_4/go/solution.go
func checkIfPrerequisite(numCourses int, prerequisites [][]int, queries [][]int) []bool {
	adj := make([]map[int]bool, numCourses)
	for i := range adj {
		adj[i] = make(map[int]bool)
	}

	indegree := make([]int, numCourses)
	isPrereq := make([]map[int]bool, numCourses)
	for i := range isPrereq {
		isPrereq[i] = make(map[int]bool)
	}

	for _, edge := range prerequisites {
		pre, crs := edge[0], edge[1]
		adj[pre][crs] = true
		indegree[crs]++
	}

	q := []int{}
	for i := 0; i < numCourses; i++ {
		if indegree[i] == 0 {
			q = append(q, i)
		}
	}

	for len(q) > 0 {
		node := q[0]
		q = q[1:]

		for neighbor := range adj[node] {
			isPrereq[neighbor][node] = true
			for prereqNode := range isPrereq[node] {
				isPrereq[neighbor][prereqNode] = true
			}

			indegree[neighbor]--
			if indegree[neighbor] == 0 {
				q = append(q, neighbor)
			}
		}
	}

	result := make([]bool, len(queries))
	for i, query := range queries {
		u, v := query[0], query[1]
		result[i] = isPrereq[v][u]
	}

	return result
}
